Check ownership of config file

ArangoGutierrez opened the pull request:
**Description of the Pull Request (PR):**

Check for ownership of config file before running

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](../CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](../CHANGELOG.md) if necessary according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](../CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](../CONTRIBUTORS.md)


Attn: @singularity-maintainers

ArangoGutierrez committed:
Check ownership of config file

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

ArangoGutierrez committed:
use sylog instead of log

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

jmstover reviewed (approved):
Works for me. I'll let @cclerget and/or @tri-adam speak up here.

cclerget commented on internal/app/singularity/actions.go:
```diff
@@ -310,6 +310,14 @@ func execStarter(cobraCmd *cobra.Command, image string, args []string, name stri
 	engineConfig := singularity.NewConfig()
 
 	configurationFile := buildcfg.SYSCONFDIR + "/singularity/singularity.conf"
+	configFile, err := os.Stat(configurationFile)
```
The whole check could be moved around this line : https://github.com/sylabs/singularity/blob/d61d61e2594c50ac3e1e0611c16a087c2f57d09f/internal/pkg/runtime/engines/singularity/prepare.go#L417

cclerget commented on internal/app/singularity/actions.go:
```diff
@@ -310,6 +310,14 @@ func execStarter(cobraCmd *cobra.Command, image string, args []string, name stri
+	if err != nil {
+		sylog.Fatalf("Error with Stat() on singularity.conf: %v", err)
+	}
+	if configFile.Sys().(*syscall.Stat_t).Uid != 0 {
```
Should we check group too ?

jmstover replied:
Only ownership was checked in 2.x, and we're replicating that.  `src/util/suid.c` ~line 75

GodloveD replied:
Well, just because we were only checking user ownership in 2.x doesn't mean we should only check the same here.  Seems sensible to check the group.  And probably not hard either right?

jmstover replied:
As I told @ArangoGutierrez , hard-coding a group check for root will not make you fans with folks that change the perms like:
```
chmod 664 $sysconfdir/singularity/singularity.conf
chown root:singularity $sysconfdir/singularity/singularity.conf
```
So, the _group_ **singularity** can edit the `singularity.conf` file. Ownership of the file fine, but not group.

cclerget reviewed (dismissed):
Need to move the code in `internal/pkg/runtime/engines/singularity/prepare.go` and execute check when `starterConfig.GetIsSUID()` returns true. Maybe consider additional checks on owner/group of `singularity` configuration directory, `capabilities` and `ecl.toml`

ArangoGutierrez committed:
Attend Cedric's comment
move the code in actions.go to internal/pkg/runtime/engines/singularity/prepare.go and execute check when starterConfig.GetIsSUID() returns true.
Add additional checks on owner of capabilities and ecl.toml files

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

ArangoGutierrez committed:
Uid -> UID to satisfy lint

Signed-off-by: ArangoGutierrez <eduardo@sylabs.io>

bauerm97 commented on internal/pkg/runtime/engines/config/check.go:
```diff
@@ -0,0 +1,25 @@
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package config
```
I don't think `internal/pkg/runtime/engines/config` is the right package for a utility function which is just checking the file ownership.

bauerm97 commented on pkg/util/capabilities/file.go:
```diff
@@ -11,6 +11,8 @@ import (
 	"os"
 	"syscall"
+
+	"github.com/sylabs/singularity/internal/pkg/runtime/engines/config"
```
As I mentioned in the other comment, we shouldn't be importing `internal/pkg/runtime/...` from a package in `pkg/...`. This would (likely) break any external thing importing this package

jscook2345 commented on internal/pkg/runtime/engines/config/check.go:
```diff
@@ -0,0 +1,25 @@
+
+// CheckUID for ownership of config file before reading it
+func CheckUID(filepath string) error {
+	fmt.Println(filepath)
```
Remove please.

ArangoGutierrez committed:
Use IsOwner() func

Signed-off-by: Eduardo Arango <eduardo@sylabs.io>

ArangoGutierrez committed:
Merge branch 'master' into check_ownership_config
